source/agora/node/Ledger.d

```diff
@@ -253,10 +253,12 @@ public class Ledger
         bool validators_changed = block.header.enrollments.length > 0
             || this.enroll_man.validatorCount() != old_count;
 
+        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
```

Comment:
This should all be moved inside of `EnrollmentManager`. Because now we have enrollment logic living in both here and also in the EnrollmentManager.

Also what does the Ledger know about the node? It doesn't know whether it's a FullNode or a Validator, so this `getEnrollmentPublicKey` call is out of place here..

Here's one way to improve the code:

```diff
diff --git a/source/agora/consensus/EnrollmentManager.d b/source/agora/consensus/EnrollmentManager.d
index b60ac75d0..1619e824b 100644
--- a/source/agora/consensus/EnrollmentManager.d
+++ b/source/agora/consensus/EnrollmentManager.d
@@ -208,15 +208,17 @@ public class EnrollmentManager
     ***************************************************************************/
 
     public string addValidator (const ref Enrollment enroll,
-        Height block_height, scope UTXOFinder finder,
-        const ref UTXOSetValue[Hash] self_utxos) @safe nothrow
+        Height block_height, scope UTXOFinder finder) @safe nothrow
     {
         this.enroll_pool.remove(enroll.utxo_key);
 
         if (auto r = this.validator_set.add(block_height, finder, enroll))
             return r;
 
-        if (enroll.utxo_key in self_utxos)
+        const pubkey = this.getEnrollmentPublicKey();
+        UTXOSetValue utxo;
+        if (finder(enroll.utxo_key, size_t.max, utxo) &&
+            utxo.output.address == pubkey)
         {
             this.data = enroll;
             this.enroll_key = enroll.utxo_key;
@@ -479,7 +481,7 @@ public class EnrollmentManager
     ***************************************************************************/
 
     public void restoreValidators (Height last_height, const ref Block block,
-        scope UTXOFinder finder, const ref UTXOSetValue[Hash] self_utxos)
+        scope UTXOFinder finder)
         @safe nothrow
     {
         assert(last_height >= block.header.height);
@@ -487,8 +489,7 @@ public class EnrollmentManager
         {
             foreach (const ref enroll; block.header.enrollments)
             {
-                this.addValidator(enroll, block.header.height, finder,
-                    self_utxos);
+                this.addValidator(enroll, block.header.height, finder);
             }
         }
     }
@@ -791,9 +792,9 @@ unittest
 
     // test for enrollment block height update
     assert(man.getEnrolledHeight(utxo_hash) == ulong.max);
-    assert(man.addValidator(enroll, Height(9), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll, Height(9), &storage.findUTXO) is null);
     assert(man.getEnrolledHeight(enroll.utxo_key) == 9);
-    assert(man.addValidator(enroll, Height(9), &storage.findUTXO, utxos) !is null);
+    assert(man.addValidator(enroll, Height(9), &storage.findUTXO) !is null);
     assert(man.getEnrolledHeight(enroll2.utxo_key) == ulong.max);
     man.pool.getEnrollments(enrolls);
     assert(enrolls.length == 1);
@@ -826,7 +827,7 @@ unittest
     // So, a pre-image can only be got from the start height.
     PreImageInfo preimage;
     assert(man.createEnrollment(utxo_hash, Height(1), enroll));
-    assert(man.addValidator(enroll, Height(10), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll, Height(10), &storage.findUTXO) is null);
     assert(!man.getPreimage(Height(10), preimage));
     assert(man.getPreimage(Height(11), preimage));
     assert(preimage.hash == man.cycle.preimages[$ - 1]);
@@ -850,7 +851,7 @@ unittest
     // validator A with the `utxo_hash` and the enrolled height of 10.
     // validator B with the 'utxo_hash2' and the enrolled height of 11.
     // validator C with the 'utxo_hash3' and no enrolled height.
-    assert(man.addValidator(enroll2, Height(11), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll2, Height(11), &storage.findUTXO) is null);
     man.clearExpiredValidators(Height(11));
     assert(man.validatorCount() == 2);
     assert(man.getEnrolledUTXOs(keys));
@@ -859,7 +860,7 @@ unittest
     // set an enrolled height for validator C
     // set the block height to 1019, which means validator B is expired.
     // there is only one validator in the middle of 1020th block being made.
-    assert(man.addValidator(enroll3, Height(1019), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll3, Height(1019), &storage.findUTXO) is null);
     man.clearExpiredValidators(Height(1019));
     assert(man.validatorCount() == 1);
     assert(man.getEnrolledUTXOs(keys));
@@ -911,7 +912,7 @@ unittest
     assert(man.params.ValidatorCycle - 101 == 907); // Sanity check
     assert(man.getValidatorPreimage(utxo_hash) == PreImageInfo.init);
     auto preimage = PreImageInfo(utxo_hash, man.cycle.preimages[100], 907);
-    assert(man.addValidator(enroll, Height(2), &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enroll, Height(2), &storage.findUTXO) is null);
     assert(man.addPreimage(preimage));
     assert(man.getValidatorPreimage(utxo_hash) == preimage);
 }
@@ -1033,7 +1034,7 @@ unittest
     assert(man.getValidatorCount(block_height) + 1 == 1);
 
     man.clearExpiredValidators(block_height);
-    assert(man.addValidator(enrollment, block_height, &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enrollment, block_height, &storage.findUTXO) is null);
 
     block_height = 3;
 
@@ -1044,7 +1045,7 @@ unittest
     assert(man.getValidatorCount(block_height) + 1 == 2);
 
     man.clearExpiredValidators(block_height);
-    assert(man.addValidator(enrollment, block_height, &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enrollment, block_height, &storage.findUTXO) is null);
 
     block_height = 4;
 
@@ -1055,7 +1056,7 @@ unittest
     assert(man.getValidatorCount(block_height) + 1 == 3);
 
     man.clearExpiredValidators(block_height);
-    assert(man.addValidator(enrollment, block_height, &storage.findUTXO, utxos) is null);
+    assert(man.addValidator(enrollment, block_height, &storage.findUTXO) is null);
 
     block_height = 5;    // valid block height : 0 <= H < 1008
     man.clearExpiredValidators(block_height);
@@ -1122,13 +1123,13 @@ unittest
     // Now create our own Enrollment
     // This means adding enrollment data to the block at the height of 2, 5, 7
     assert(man.createEnrollment(utxo_keys[0], Height(1), enroll));
-    man.addValidator(enroll, Height(2), &utxo_set.findUTXO, utxos);
+    man.addValidator(enroll, Height(2), &utxo_set.findUTXO);
     enrolls ~= enroll;
     assert(man.createEnrollment(utxo_keys[1], Height(2), enroll));
-    man.addValidator(enroll, Height(5), &utxo_set.findUTXO, utxos);
+    man.addValidator(enroll, Height(5), &utxo_set.findUTXO);
     enrolls ~= enroll;
     assert(man.createEnrollment(utxo_keys[2], Height(3), enroll));
-    man.addValidator(enroll, Height(7), &utxo_set.findUTXO, utxos);
+    man.addValidator(enroll, Height(7), &utxo_set.findUTXO);
     enrolls ~= enroll;
 
     PreImageInfo org_preimage;
@@ -1136,9 +1137,9 @@ unittest
 
     // clear up the EnrollmentManager for test
     man = new EnrollmentManager(":memory:", gen_key_pair, params);
-    man.addValidator(enrolls[0], Height(2), &utxo_set.findUTXO, utxos);
-    man.addValidator(enrolls[1], Height(5), &utxo_set.findUTXO, utxos);
-    man.addValidator(enrolls[2], Height(7), &utxo_set.findUTXO, utxos);
+    man.addValidator(enrolls[0], Height(2), &utxo_set.findUTXO);
+    man.addValidator(enrolls[1], Height(5), &utxo_set.findUTXO);
+    man.addValidator(enrolls[2], Height(7), &utxo_set.findUTXO);
 
     // check the restored enrollment data
     assert(man.enroll_key == enrolls[2].utxo_key);
diff --git a/source/agora/node/Ledger.d b/source/agora/node/Ledger.d
index 0f55bd7ee..78a140f0c 100644
--- a/source/agora/node/Ledger.d
+++ b/source/agora/node/Ledger.d
@@ -137,14 +137,12 @@ public class Ledger
 
         // restore validator set from the blockchain.
         // using block_count, as the range is inclusive
-        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
         foreach (block_idx; min_height .. block_count)
         {
             Block block;
             this.storage.readBlock(block, block_idx);
-            UTXOSetValue[Hash] utxos = this.utxo_set.getUTXOs(pubkey);
             this.enroll_man.restoreValidators(this.last_block.header.height,
-                block, this.utxo_set.getUTXOFinder(), utxos);
+                block, this.utxo_set.getUTXOFinder());
         }
     }
 
@@ -255,12 +253,10 @@ public class Ledger
         bool validators_changed = block.header.enrollments.length > 0
             || this.enroll_man.validatorCount() != old_count;
 
-        PublicKey pubkey = this.enroll_man.getEnrollmentPublicKey();
-        UTXOSetValue[Hash] utxos = this.utxo_set.getUTXOs(pubkey);
         foreach (idx, ref enrollment; block.header.enrollments)
         {
             if (auto r = this.enroll_man.addValidator(enrollment,
-                block.header.height, this.utxo_set.getUTXOFinder(), utxos))
+                block.header.height, this.utxo_set.getUTXOFinder()))
             {
                 log.fatal("Error while adding a new validator: {}", r);
                 log.fatal("Enrollment #{}: {}", idx, enrollment);
```

Reply:
> Here's one way to improve the code:
```
-        if (enroll.utxo_key in self_utxos)
+        const pubkey = this.getEnrollmentPublicKey();
+        UTXOSetValue utxo;
+        if (finder(enroll.utxo_key, size_t.max, utxo) &&
+            utxo.output.address == pubkey)
```
The suggestion by @AndrejMitrovic didn't work for my revised network test which is in the last commit. So, I have returned to my original code about that.

Reply:
Why? What was failing?

Reply:
The code for adding validators has failed with the `finder`.  You can check that with my last commit having the network test. I guess it's because the `finder` function put a UTXOSetValue passed through a parameter into the `used_utxos'.